src/main/java/org/folio/rest/impl/HoldingsStorageAPI.java

```diff
@@ -408,4 +431,60 @@ private boolean isUUID(String id) {
       return false;
     }
   }
+
+  private CompletableFuture<Void> updateItemEffectiveCallNumbersByHoldings(AsyncResult<SQLConnection> connection, PostgresClient postgresClient, HoldingsRecord holdingsRecord) {
+    CompletableFuture<Void> future = new CompletableFuture<>();
+    Criterion criterion = new Criterion(
+      new Criteria().addField("holdingsRecordId")
+        .setJSONB(false).setOperation("=").setVal(holdingsRecord.getId()));
+    postgresClient.get(ITEM_TABLE, Item.class, criterion, false, false, response -> {
+      updateEffectiveCallNumbers(connection, postgresClient, response.result().getResults(), holdingsRecord).thenAccept(voidResult -> {
+        future.complete(null);
+      });
+    });
+    return future;
+  }
+
+  private CompletableFuture<Void> updateEffectiveCallNumbers(AsyncResult<SQLConnection> connection, PostgresClient postgresClient, List<Item> items, HoldingsRecord holdingsRecord) {
+    CompletableFuture<Void> allItemsUpdated = new CompletableFuture<>();
+    List<Function<SQLConnection, Future<UpdateResult>>> batchFactories = items
+      .stream()
+      .map(item -> updateItemEffectiveCallNumber(item, holdingsRecord))
+      .map(item -> updateSingleBatchFactory(ITEM_TABLE, item.getId(), item, postgresClient))
+      .collect(Collectors.toList());
+
+    SQLConnection connectionResult = connection.result();
+    Future<UpdateResult> lastUpdate = Future.succeededFuture();
+    for (Function<SQLConnection, Future<UpdateResult>> factory : batchFactories) {
+      lastUpdate = lastUpdate.compose(prev -> factory.apply(connectionResult));
+    }
+
+    lastUpdate.setHandler(updateResult -> {
+      if (updateResult.succeeded()) {
```

Comment:
I think we have to somehow notify calling method about failures (if any) so the transaction will be reverted if an item update has failed. You can use `Future` for this, and handle overall result in handler (`setHandler` method). 

Alternatively, you can use following algorithm:
1. Build a list of update factories (`updateSingleBatchFactory` method);
    1. The first part of the list will be items, that needs to be updated;
    1. The last element will be the holding itself.
1. Then, you will need a single handler for the last `Future` which will decide whether to commit or revert transaction.

The code can be something like:

```java
  private void updateHoldingAndItems(
    PostgresClient postgresClient, HoldingsRecord oldHolding, HoldingsRecord newHolding, 
    Handler<AsyncResult<Void>> handler) {

    if (Objects.equals(oldHolding.getCallNumber(), newHolding.getCallNumber())) {
      // Call number has not been changed
      // just do the holding simple update
      postgresClient.update(HOLDINGS_RECORD_TABLE, newHolding, newHolding.getId(), 
          result -> {/* handle result */});
      return;
    }

    // Call number has changed.
    final List<Item> itemsForHolding = getItemsForHolding(newHolding.getId());
    postgresClient.startTx(connectionResult -> {
      List<Function<SQLConnection, Future<UpdateResult>>> itemsWithNewCallNumber = ...
          /** update effective call number, 
            * remove those for which effective call number has not changed
            * wrap into the function
            */

      List<Function<SQLConnection, Future<UpdateResult>>> allUpdates = 
          new ArrayList<>(itemsWithNewCallNumber);
      allUpdates.add(updateSingleBatchFactory(HOLDINGS_RECORD_TABLE, newHolding.getId(), 
          newHolding, postgresClient));
      
      Future<UpdateResult> lastUpdate = Future.succeededFuture();
      for (Function<SQLConnection, Future<UpdateResult>> update : allUpdates) {
        lastUpdate = lastUpdate
          .compose(prev -> update.apply(connectionResult.result()));
      }
      
      lastUpdate.setHandler(updateResult -> {/* do commit or revert */ });
    });
  }
```
 
Then, you can execute this method in `putHoldingsStorageHoldingsByHoldingsRecordId`, handler will just return 204 or 500, as usual.
Note:  `oldHolding` above can be retrieved from the `holdingsList`, that we fetch from db in `putHoldingsStorageHoldingsByHoldingsRecordId`.

Reply:
Is it possible to start a new transaction within an existing transaction?

The alternate approach does seem to have some advantages over the current one, but I don't want to delay this PR any more than necessary. Would it be an acceptable compromise to create an issue with the alternate algorithm information that I could get back to if I have time after the call number prefix and suffix work is done?

Reply:
I would think all you need to do is
```suggestion
      if (updateResult.succeeded()) {
        allItemsUpdated.complete(null);
      } else {
        allItemsUpdated.fail(updateResult.cause());
      }
```

Since `compose` will not call its mapper if the future it is called on has failed, the `lastUpdate` will be the failed future with its reason. You'd have to fail it up the chain since you are returning different futures in each method. I think that is a good idea anyway since I think the way it is currently implemented could lead to deadlock if there are failures (if updateResult did not succeed, allItemsUpdate is never completed).

Reply:
Sure, @rladdusaw, you can use another logic here. My main concern that we don't handle failures for item updates, we have to somehow propagate failure, so we will revert the transaction. 

As an option: change `allItemsUpdated` to `Future` and check whether it failed or completed in `updateItemEffectiveCallNumbersByHoldings` or in `putHoldingsStorageHoldingsByHoldingsRecordId` methods.

I've pushed a holdings update service https://github.com/folio-org/mod-inventory-storage/compare/MODINVSTOR-357-effective-call-number...MODINVSTOR-357-effective-call-number-service?expand=1, hope you find it useful.